### PR TITLE
Remove redundant flags from `lower_ty_common` that can be inferred from the HIR

### PR DESCRIPTION
**Pull request by compiler-errors**
...and then get rid of `lower_ty_common`.

r? @fmease or re-roll if you're busy!

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/object_safety.rs [hunk 1]
```diff
@@ -325,22 +324,32 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
         } else {
             self.compute_object_lifetime_bound(span, existential_predicates).unwrap_or_else(|| {
+                // Curiously, we prefer object lifetime default for `+ '_`...
                 if tcx.named_bound_var(lifetime.hir_id).is_some() {
-                    self.lower_lifetime(lifetime, RegionInferReason::ObjectLifetimeDefault)
+                    self.lower_lifetime(lifetime, RegionInferReason::ExplicitObjectLifetime)
                 } else {
-                    self.re_infer(
-                        span,
-                        if borrowed {
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/object_safety.rs:338]: This was so confusing. `if borrowed`, then return `ObjectLifetimeDefault`... and `!borrowed` then return `BorrowedObjectLifetimeDefault`?! That feels like a typo.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs [hunk 2]
```diff
@@ -28,6 +24,17 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             return;
         };
 
+        let in_path = match tcx.parent_hir_node(self_ty.hir_id) {
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs:27]: I'm pretty sure this needs to handle paths in `Node::Pat`, too.

IINM, not doing so regresses the suggestion for the following code:

```rs
trait Trait {}
impl dyn Trait { const CONST: () = (); }

fn main() {
    match () { Trait::CONST => {} }
}
```

Should keep suggesting `<dyn Trait>::CONST` inside the match.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/object_safety.rs (hunk 1 above)
Comment (fmease) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/object_safety.rs:338]: Good catch, yeah I must've overlooked this when reviewing the PR that introduced the enum

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/object_safety.rs [hunk 3]
```diff
@@ -325,22 +324,32 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
         let existential_predicates = tcx.mk_poly_existential_predicates(&v);
 
-        // Use explicitly-specified region bound.
+        // Use explicitly-specified region bound, unless the bound is missing.
         let region_bound = if !lifetime.is_elided() {
-            self.lower_lifetime(lifetime, RegionInferReason::ObjectLifetimeDefault)
+            self.lower_lifetime(lifetime, RegionInferReason::ExplicitObjectLifetime)
         } else {
             self.compute_object_lifetime_bound(span, existential_predicates).unwrap_or_else(|| {
+                // Curiously, we prefer object lifetime default for `+ '_`...
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/object_safety.rs:332]: Curious indeed. I'll investigate this at some point

**Review by fmease — APPROVED** [review]

**Comment by fmease** [comment]
Fabulous! That was quite a wart. r=me with nit addressed
@bors rollup

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs:27]: Thanks for catching that! I love how we didn't have a test for that 💀

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/object_safety.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/object_safety.rs:332]: I think ironically that it's beneficial -- I tried preferring the explicit elided lifetime and it causes some outlives bounds to fail.

**Pull request merged**